read-fonts/src/tables/truetype/deltas.rs

```diff
@@ -0,0 +1,350 @@
+use core::ops::RangeInclusive;
+
+use crate::{
+    tables::glyf::{PointFlags, PointMarker},
+    tables::gvar::{Gvar, TupleVariation},
+    types::{F2Dot14, Fixed, GlyphId, Point},
+    ReadError,
+};
+
+pub type Delta = Point<Fixed>;
+
+/// Compute a set of deltas for the component offsets of a composite glyph.
+///
+/// Interpolation is meaningless for component offsets so this is a
+/// specialized function that skips the expensive bits.
+pub fn composite_glyph(
+    gvar: &Gvar,
+    glyph_id: GlyphId,
+    coords: &[F2Dot14],
+    deltas: &mut [Delta],
+) -> Result<(), ReadError> {
+    compute_deltas_for_glyph(gvar, glyph_id, coords, deltas, |scalar, tuple, deltas| {
+        for tuple_delta in tuple.deltas() {
+            let ix = tuple_delta.position as usize;
+            if let Some(delta) = deltas.get_mut(ix) {
+                *delta += tuple_delta.apply_scalar(scalar);
+            }
+        }
+        Ok(())
+    })?;
+    Ok(())
+}
+
+pub struct SimpleGlyph<'a> {
+    pub points: &'a [Point<i32>],
+    pub flags: &'a mut [PointFlags],
+    pub contours: &'a [u16],
+}
+
+/// Compute a set of deltas for the points in a simple glyph.
+///
+/// This function will use interpolation to infer missing deltas for tuples
+/// that contain sparse sets. The `iup_buffer` buffer is temporary storage
+/// used for this and the length must be >= glyph.points.len().
+pub fn simple_glyph(
+    gvar: &Gvar,
+    glyph_id: GlyphId,
+    coords: &[F2Dot14],
+    has_var_lsb: bool,
+    glyph: SimpleGlyph,
+    iup_buffer: &mut [Point<Fixed>],
+    deltas: &mut [Delta],
+) -> Result<(), ReadError> {
+    if iup_buffer.len() < glyph.points.len() {
+        return Err(ReadError::InvalidArrayLen);
+    }
+    for delta in deltas.iter_mut() {
+        *delta = Default::default();
+    }
+    if gvar.glyph_variation_data(glyph_id).is_err() {
+        // Empty variation data for a glyph is not an error.
+        return Ok(());
+    };
+    let SimpleGlyph {
+        points,
+        flags,
+        contours,
+    } = glyph;
+    // Include the first phantom point if the font is missing variable metrics
+    // for left side bearings. The adjustment made here may affect the final
+    // shift of the outline.
+    let actual_len = if has_var_lsb {
+        points.len() - 4
```

Comment:
is it possible for `points` to be shorter than expected?

Reply:
We always push phantom points before calling into this so it should be at least 4, but yes, if we make this public, we’ll need an extra check. I’ll add one to be safe.

Reply:
Done